Range type.

paulp opened the pull request:
Go slices fail hard if the inputs don't make sense, but
slices are much more convenient to work with if they do the
sensible thing when you overshoot, e.g. slice[0:10] if there
are only 5 elements would give you the 5. Range superimposes
the scala-like forgiving slice logic on top of native slices.

paulp committed:
Range type.

Go slices fail hard if the inputs don't make sense, but
slices are much more convenient to work with if they do the
sensible thing when you overshoot, e.g. slice[0:10] if there
are only 5 elements would give you the 5. Range superimposes
the scala-like forgiving slice logic on top of native slices.

EECOLOR commented on suffuse/range.go:
```diff
@@ -0,0 +1,79 @@
+var EmptyRange Range = Range { Offset(0), Length(0) }
+
+func (x Range) IsEmpty()bool  { return x.LengthInt() == 0           }
+func (x Range) StartInt()int  { return int(x.Offset)                }
```
Any reason this is called `StartInt` instead of `Start`?

paulp replied:
Because Start is an Offset. I avoid returning types-which-aren't-types like int wherever possible, but I generally embed the name of the primitive type when I do so to highlight the fact.

EECOLOR commented on suffuse/range.go:
```diff
@@ -0,0 +1,79 @@
+  if offset < 0 {
+    return RangeOffsetLength(0, length)
+  } else if length <= 0 {
+    return EmptyRange
```
This ignores the offset. Theoretically that might be a problem later on when ranges can be expanded.

paulp replied:
There's a comment:

```
+/** All empty ranges are defined as 0:0. We don't want the
+ *  start value of an empty range to be considered meaningful
+ *  or stable under any operations.
+ */
```

EECOLOR replied:
Ahh, good, missed that one

EECOLOR commented on suffuse/range.go:
```diff
@@ -0,0 +1,79 @@
+  if start < 0 {
+    return RangeStartEnd(0, end)
+  } else if end <= start || end <= 0 {
+    return EmptyRange
```
Same thing here when `end == start`, not sure if that will ever be a problem though

paulp replied:
Same comment applies. It's intentional.

EECOLOR commented:
Looks great!

Do you mind me changing the following two things when merging?
1. Separate the tests into smaller methods in order to remove the comments
2. Adding spaces between the function and return types

EECOLOR commented:
After performing the change I am not convinced about 1

paulp commented:
I also am doubtful about 1. It's a very brute force test. I wouldn't test most of it if there was going to be more overhead.

paulp commented:
I _could_ add space between the return types, but I hate to set the precedent that we should be fiddling with exact spacing. If it's important let's write our own go formatter (or just use go fmt, as unbelievably hideous as its output is), if its isn't let's forget it.

But yeah, I'd rather use `go fmt` then spend bandwidth on the presence or absence of a space.

EECOLOR committed:
Moved some code and introduced spaces

- Moved `min` method to bottom of file
- Added spaces before return types
- Added blank lines to group methods
- Moved tests for byte slices to a separate file

EECOLOR commented:
> I could add space between the return types, but I hate to set the precedent that we should be fiddling with exact spacing.

That's fine. If you don't care too much it means I can just change them without running into trouble. They bug me, haha. And writing a formatter is off-track from where we want to go.

I added a commit with proposed changes.
